File: Arrays/searching-algorithms/block-search/block.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

class BlockSearch {
    std::vector<int> arr;
    std::vector<int> indexTable;
    int blockSize;
    int numBlocks;

public:
    explicit BlockSearch(const std::vector<int>& inputArr) : arr(inputArr) {
        int n = arr.size();
        if (n == 0) return;

        blockSize = static_cast<int>(std::sqrt(n));
        numBlocks = (n + blockSize - 1) / blockSize;

        // Build index table: stores the first element of each block
        indexTable.reserve(numBlocks);
        for (int i = 0; i < n; i += blockSize) {
            indexTable.push_back(arr[i]);
        }
    }

    int search(int target) const {
        if (arr.empty()) return -1;

        // 1. Find correct block using index table
        int blockIdx = 0;
        while (blockIdx < numBlocks && indexTable[blockIdx] <= target) {
            blockIdx++;
        }
        blockIdx--; // Step back to the block that might contain target

        // 2. Define search bounds within the identified block
        int start = blockIdx * blockSize;
        int end = std::min(start + blockSize, static_cast<int>(arr.size()));

        // 3. Linear search within the block
        for (int i = start; i < end; ++i) {
            if (arr[i] == target) return i;
        }
        return -1;
    }
};

int main() {
    std::vector<int> arr = {5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65};
    BlockSearch bs(arr);

    std::cout << "Index of 40: " << bs.search(40) << "\n";
    std::cout << "Index of 99: " << bs.search(99) << "\n";

    return 0;
}

File: Arrays/searching-algorithms/block-search/block.js
/**
 * Block Search (Indexed Sequential Search)
 * Preprocesses array into blocks with an index table for efficient repeated queries.
 */
class BlockSearch {
    #arr = [];
    #indexTable = [];
    #blockSize = 0;
    #numBlocks = 0;

    /**
     * @param {number[]} inputArr
     */
    constructor(inputArr) {
        this.#arr = inputArr;
        const n = inputArr.length;
        if (n === 0) return;

        this.#blockSize = Math.floor(Math.sqrt(n));
        this.#numBlocks = Math.ceil(n / this.#blockSize);

        // Build index table: stores first element of each block
        for (let i = 0; i < n; i += this.#blockSize) {
            this.#indexTable.push(inputArr[i]);
        }
    }

    /**
     * @param {number} target
     * @returns {number} Index or -1
     */
    search(target) {
        if (this.#arr.length === 0) return -1;

        // 1. Find correct block
        let blockIdx = 0;
        while (blockIdx < this.#indexTable.length && this.#indexTable[blockIdx] <= target) {
            blockIdx++;
        }
        blockIdx--;

        // 2. Search within block
        const start = blockIdx * this.#blockSize;
        const end = Math.min(start + this.#blockSize, this.#arr.length);

        for (let i = start; i < end; i++) {
            if (this.#arr[i] === target) return i;
        }
        return -1;
    }
}

// Example Usage
const arr = [5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65];
const bs = new BlockSearch(arr);

console.log(`Index of 40: ${bs.search(40)}`);
console.log(`Index of 99: ${bs.search(99)}`);